core/feed_protocol/types.py

```diff
@@ -0,0 +1,160 @@
+from typing import Any, Dict, Generator, List, Optional, Tuple
+
+from pydantic import BaseModel
+
+from core.model import LicensePool, Work
+from core.model.edition import Edition
+from core.model.identifier import Identifier
+
+
+class FeedEntryType(BaseModel):
+    text: Optional[str] = None
+
+    class Config:
+        extra = "allow"
+        arbitrary_types_allowed = True
+
+    def add_attributes(self, attrs: Dict[str, Any]) -> None:
+        for name, data in attrs.items():
+            setattr(self, name, data)
+
+    def children(self) -> Generator[Tuple[str, "FeedEntryType"], None, None]:
+        """Yield all FeedEntryType attributes"""
+        for name, value in self:
+            if isinstance(value, self.__class__):
+                yield name, value
+        return
+
+
+class Link(FeedEntryType):
+    href: str
+    rel: Optional[str]
+    type: Optional[str]
+
+    # Additional types
+    role: Optional[str] = None
+    title: Optional[str] = None
+
+    def dict(self, **kwargs: Any) -> Dict[str, Any]:
+        kwargs["exclude_none"] = True
+        return super().dict(**kwargs)
+
+    def link_attribs(self) -> Dict[str, Any]:
+        d = dict(href=self.href)
+        for key in ["rel", "type"]:
+            if (value := getattr(self, key, None)) is not None:
+                d[key] = value
+        return d
+
+
+class IndirectAcquisition(BaseModel):
+    type: str
+    children: List["IndirectAcquisition"] = []
+
+
+class Acquisition(Link):
+    holds_position: Optional[str]
+    holds_total: Optional[str]
+
+    copies_available: Optional[str]
+    copies_total: Optional[str]
+
+    availability_status: Optional[str]
+    availability_since: Optional[str]
+    availability_until: Optional[str]
+
+    rights: Optional[str]
+
+    lcp_hashed_passphrase: Optional[FeedEntryType]
+    drm_licensor: Optional[FeedEntryType]
+
+    indirect_acquisitions: List[IndirectAcquisition] = []
+
+
+class Author(FeedEntryType):
+    name: Optional[str]
+    sort_name: Optional[str]
+    viaf: Optional[str]
+    role: Optional[str]
+    family_name: Optional[str]
+    wikipedia_name: Optional[str]
+    lc: Optional[str]
+    link: Optional[Link]
+
+
+class WorkEntryData(BaseModel):
```

Comment:
Should this represent the OPDS 2.x metadata collection?
https://drafts.opds.io/opds-2.0.html#42-metadata

Reply:
In #1340 the serializer formats the output as required.